File: bfinter.lua
tmp = io.read("*l")
program = {}
for i=1, string.len(tmp)  do    --Lua is 1-indexed.
    program[i] = string.char(tmp:byte(i))
end
memorysize = 65536; --64KB
pointer = 1;
i = 1;
bra = 0;
arr = {};
for tmp = 0, memorysize-1  do
    table.insert(arr, 0)
end
while i <= #program do
    if program[i] == '+' then
        if 0 <= pointer and pointer < memorysize then
            arr[pointer] = arr[pointer] + 1;
        else
            print("Out Of Array Bounds At: " + i);
            return;
        end
        i = i + 1;
    elseif program[i] == '-' then
        if 0 <= pointer and pointer < memorysize then
            arr[pointer] = arr[pointer] - 1;
        else
            print("Out Of Array Bounds At: " + i);
            return;
        end
        i = i + 1;
    elseif program[i] == '<' then
        pointer = pointer - 1
        i = i + 1;
    elseif program[i] == '>' then
        pointer = pointer + 1
        i = i + 1;
    elseif program[i] == '.' then
        --print(arr[pointer])
        io.write(string.char(arr[pointer]))
        i = i + 1;
    elseif program[i] == ',' then
        inp = io.read("*n")
        arr[pointer] = inp
        i = i + 1
    elseif program[i] == '[' then
        if arr[pointer] == 0 then
            bra = 1;
            while bra > 0 do
                i = i + 1;
                if i <= #program and program[i] == '[' then
                    bra = bra + 1;
                elseif i < #program and program[i] == ']' then
                    bra = bra - 1;
                elseif i >= #program then
                    print("EOF Reached");
                end
            end
        else
            i = i + 1;
        end
    elseif program[i] == ']' then
        if arr[pointer] ~= 0 then
            bra = 1;
            while bra > 0 do
                i = i - 1;
                if i >= 0 and program[i] == '[' then
                    bra = bra - 1;
                elseif i >= 0 and program[i] == ']' then
                    bra = bra + 1;
                elseif i < 0 then
                    print("Out Of Program.");
                    return
                end
            end
        else
            i = i + 1;
        end
    else
        i = i + 1;
    end
end

File: bfinter.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	program, _ := reader.ReadString('\n')
	memorysize := 65536
	pointer := 0
	arr := make([]int, memorysize)
	bra := 0
	i := 0
	for i < len(program) {
		if program[i] == '+' {
			if 0 <= pointer && pointer < memorysize {
				arr[pointer]++
			} else {
				fmt.Printf("Out Of Array Bounds At: %v\n", i)
			}
			i++
		} else if program[i] == '-' {
			if 0 <= pointer && pointer < memorysize {
				arr[pointer]--
			} else {
				fmt.Printf("Out Of Array Bounds At: %v\n", i)
			}
			i++
		} else if program[i] == '<' {
			pointer--
			i++
		} else if program[i] == '>' {
			pointer++
			i++
		} else if program[i] == ',' {
			tmp, _ := reader.ReadString('\n')
			inp, _ := strconv.Atoi(tmp)
			arr[pointer] = inp
			i++
		} else if program[i] == '.' {
			fmt.Printf("%c", arr[pointer])
			i++
		} else if program[i] == '[' {
			if arr[pointer] == 0 {
				bra = 1
				for bra > 0 {
					i++
					if i < len(program) && program[i] == '[' {
						bra++
					} else if i < len(program) && program[i] == ']' {
						bra--
					} else if i >= len(program) {
						fmt.Println("EOF Reached")
						return
					} else {
					}
				}
			} else {
				i++
			}
		} else if program[i] == ']' {
			if arr[pointer] != 0 {
				bra = 1
				for bra > 0 {
					i--
					if i >= 0 && program[i] == '[' {
						bra--
					} else if i >= 0 && program[i] == ']' {
						bra--
					} else if i < 0 {
						fmt.Println("Out Of Program.")
						return
					} else {
					}
				}
			} else {
				i++
			}
		} else {
			i++
		}
	}
}
